pkg/syncer/resourcesync/controller.go

```diff
@@ -0,0 +1,348 @@
+	ctx = klog.NewContext(ctx, logger)
+	logger.Info("Starting controller")
+	defer logger.Info("Shutting down controller")
+
+	for i := 0; i < numThreads; i++ {
+		go wait.UntilWithContext(ctx, c.startWorker, time.Second)
+	}
+
+	<-ctx.Done()
+}
+
+func (c *Controller) startWorker(ctx context.Context) {
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+func (c *Controller) processNextWorkItem(ctx context.Context) bool {
+	// Wait until there is a new item in the working queue
+	k, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	key := k.(string)
+
+	// No matter what, tell the queue we're done with this key, to unblock
+	// other workers.
+	defer c.queue.Done(key)
+
+	if err := c.process(ctx, key); err != nil {
+		runtime.HandleError(fmt.Errorf("failed to sync %q: %w", key, err))
+		c.queue.AddRateLimited(key)
+		return true
+	}
+
+	c.queue.Forget(key)
+	return true
+}
+
+func (c *Controller) process(ctx context.Context, key string) error {
+	syncTarget, err := c.syncTargetLister.Get(key)
+	if apierrors.IsNotFound(err) {
+		c.stopSyncers(ctx, map[schema.GroupVersionResource]bool{})
+	}
+
+	if err != nil {
+		return err
+	}
+
+	if syncTarget.GetUID() != c.syncTargetUID {
+		return nil
+	}
+
+	requiredGVRs, err := getAllGVRs(syncTarget)
+	if err != nil {
+		return err
+	}
+
+	errs := []error{}
+	for gvr := range requiredGVRs {
+		err := c.startSyncer(ctx, gvr, syncTarget)
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	c.stopSyncers(ctx, requiredGVRs)
+
+	return errors.NewAggregate(errs)
+}
+
+// stopSyncer stop syncers for gvrs not in requiredGVRs
+func (c *Controller) stopSyncers(ctx context.Context, requiredGVRs map[schema.GroupVersionResource]bool) {
+	logger := klog.FromContext(ctx)
+
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	syncersToDelete := []schema.GroupVersionResource{}
+	for gvr := range c.syncerMap {
+		if _, ok := requiredGVRs[gvr]; !ok {
+			syncersToDelete = append(syncersToDelete, gvr)
+		}
+	}
+
+	for _, gvr := range syncersToDelete {
+		logger.WithValues("gvr", gvr.String()).V(2).Info("Stop syncer for gvr")
+		if cancel, ok := c.syncerMap[gvr]; ok {
+			cancel()
+			delete(c.syncerMap, gvr)
+		}
+	}
+}
+
+func (c *Controller) startSyncer(ctx context.Context, gvr schema.GroupVersionResource, syncTarget *workloadv1alpha1.SyncTarget) error {
+	logger := klog.FromContext(ctx)
+
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	if _, ok := c.syncerMap[gvr]; ok {
+		return nil
+	}
+
+	if len(syncTarget.Status.VirtualWorkspaces) == 0 {
+		return nil
+	}
+
+	syncerVirtualWorkspaceURL := syncTarget.Status.VirtualWorkspaces[0].URL
+
+	upstreamURL, err := url.Parse(syncerVirtualWorkspaceURL)
+	if err != nil {
+		return err
+	}
+
+	// Check whether we're in the Advanced Scheduling feature-gated mode.
+	advancedSchedulingEnabled := false
+	if syncTarget.GetAnnotations()[AdvancedSchedulingFeatureAnnotation] == "true" {
+		logger.WithValues("SyncTarget", c.syncTargetName).Info("Advanced Scheduling feature is enabled")
```

Comment:
nit: lower-case the field `syncTarget.Name` (does it not come from the reconcilitaiton keys?) and add this to the logger early.